medusa/helper/request_police.py

```diff
@@ -0,0 +1,356 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with Medusa. If not, see <http://www.gnu.org/licenses/>.
+# pylint:disable=too-many-lines
+"""Request Police Class for monitoring requests responses, and throttling or breaking where needed."""
+
+from ..bs4_parser import BS4Parser
+
+import errno
+import datetime
+import logging
+import requests
+import traceback
+
+from cachecontrol import CacheControlAdapter
+from cachecontrol.cache import DictCache
+
+from .. import app
+from ..common import USER_AGENT
+from common import http_code_description
+import certifi
+import cfscrape
+
+try:
+    from urllib.parse import splittype
+except ImportError:
+    from urllib2 import splittype
+
+logger = logging.getLogger(__name__)
+
+
+# Request police Exceptions.
+class RequestPoliceException(Exception):
+    """Generic Police Exception."""
+
+
+class PoliceRequestScoreExceeded(RequestPoliceException):
+    """Police Request Score Exception."""
+
+
+class PoliceResponseScoreExceeded(RequestPoliceException):
+    """Police Request Score Exception."""
+
+
+class PoliceReservedDailyExceeded(RequestPoliceException):
+    """Police Request Exception for exceeding the reserved daily search limit."""
+
+
+class RequestPoliceInvalidConfiguration(RequestPoliceException):
+    """Invalid or incomplete configuration provided."""
+
+
+def cloudflare_bypass(r, **kwargs):
+    # if not r.ok:
+    #     # Try to bypass CloudFlare's anti-bot protection
+    #     if r.status_code == 503 and r.headers.get('server') == u'cloudflare-nginx':
+    #         cf_prepped = MedusaSession._prepare_cf_req(session, req)
+    #         if cf_prepped:
+    #             # cf_resp = r.send(cf_prepped, stream=stream, verify=verify, proxies=proxies,
+    #             #                        timeout=timeout, allow_redirects=True)
+    #             # if cf_resp.ok:
+    #             #     return cf_resp
+    #
+    #     logger.debug(u'Requested url {url} returned status code {status}: {desc}'.format
+    #                  (url=resp.url, status=resp.status_code, desc=http_code_description(resp.status_code)))
+    #
+    #     if response_type and response_type != u'response':
+    #         return None
+    pass
+
+
+class MedusaSession(requests.Session):
+    def __init__(self, cache_etags=True, serializer=None, heuristic=None, police=None):
+        super(MedusaSession, self).__init__()
+        adapter = CacheControlAdapter(
+            DictCache(),
+            cache_etags=cache_etags,
+            serializer=serializer,
+            heuristic=heuristic,
+        )
+
+        self.mount('http://', adapter)
+        self.mount('https://', adapter)
+        self.cache_controller = adapter.controller
+        self.headers.update({'User-Agent': USER_AGENT, 'Accept-Encoding': 'gzip,deflate'})
+
+        # This can be used to attach a police object
+        self.police = police
+
+    @classmethod
+    def cloudflare_hook(cls, r, **kwargs):
+
```

Comment:
This was it @labrys. Hope it's worth my time! Man. Pain here, was re-setting the new cookies on a PreparedRequest object!